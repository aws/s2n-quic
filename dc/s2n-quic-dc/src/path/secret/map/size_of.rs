// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::status::IsRetired;
use s2n_quic_core::dc;
use std::{net::SocketAddr, sync::atomic::AtomicU64, time::Instant};

/// Provide an approximation of the size of Self, including any heap indirection (e.g., a vec
/// backed by a megabyte is a megabyte in `size`, not 24 bytes).
///
/// Approximation because we don't currently attempt to account for (as an example) padding. It's
/// too annoying to do that.
#[cfg_attr(not(test), allow(unused))]
pub(crate) trait SizeOf: Sized {
    fn size(&self) -> usize {
        // If we don't need drop, it's very likely that this type is fully contained in size_of
        // Self. This simplifies implementing this trait for e.g. std types.
        assert!(
            !std::mem::needs_drop::<Self>(),
            "{:?} requires custom SizeOf impl",
            std::any::type_name::<Self>()
        );
        std::mem::size_of::<Self>()
    }
}

impl SizeOf for Instant {}
impl SizeOf for u32 {}
impl SizeOf for SocketAddr {}
impl SizeOf for AtomicU64 {}

impl SizeOf for IsRetired {}
impl SizeOf for dc::ApplicationParams {}
