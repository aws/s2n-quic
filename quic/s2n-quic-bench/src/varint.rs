// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use criterion::{black_box, BenchmarkId, Criterion};
use s2n_codec::{DecoderBuffer, Encoder, EncoderBuffer, EncoderValue};
use s2n_quic_core::varint::VarInt;

pub fn benchmarks(c: &mut Criterion) {
    encode(c);
    decode(c);
}

#[inline(always)]
pub fn encode_n<const N: usize>(inputs: [VarInt; N], buffer: &mut [u8]) {
    let mut encoder = EncoderBuffer::new(buffer);
    for input in inputs {
        if input.encoding_size() > encoder.remaining_capacity() {
            break;
        }
        encoder.encode(&input);
    }
}

#[inline(never)]
pub fn encode_slice(input: VarInt, buffer: &mut [u8]) {
    encode_n([input], buffer);
}

#[inline(never)]
pub fn encode_array(input: VarInt, buffer: &mut [u8; 8]) {
    encode_n([input], buffer);
}

#[inline(never)]
pub fn encode_16_slice(input: VarInt, buffer: &mut [u8]) {
    encode_n([input; 16], buffer);
}

#[inline(never)]
pub fn encode_16_array(input: VarInt, buffer: &mut [u8; 8 * 16]) {
    encode_n([input; 16], buffer);
}

#[inline(always)]
pub fn decode_n<const N: usize>(buffer: &[u8]) -> Option<[VarInt; N]> {
    let mut buffer = DecoderBuffer::new(buffer);
    let mut out = [VarInt::default(); N];
    for out in out.iter_mut() {
        let (v, remaining) = buffer.decode::<VarInt>().ok()?;
        *out = v;
        buffer = remaining;
    }
    Some(out)
}

#[inline(never)]
pub fn decode_slice(buffer: &[u8]) -> Option<VarInt> {
    Some(decode_n::<1>(buffer)?[0])
}

#[inline(never)]
pub fn decode_array(buffer: &[u8; 8]) -> Option<VarInt> {
    Some(decode_n::<1>(buffer)?[0])
}

#[inline(never)]
pub fn decode_16_slice(buffer: &[u8]) -> Option<[VarInt; 16]> {
    decode_n(buffer)
}

#[inline(never)]
pub fn decode_16_array(buffer: &[u8; 8 * 16]) -> Option<[VarInt; 16]> {
    decode_n(buffer)
}

fn encode(c: &mut Criterion) {
    let mut group = c.benchmark_group("varint/encode");
    for i in [0, 1, 5, 6, 13, 14, 29, 30, 61] {
        let i = VarInt::new(2u64.pow(i)).unwrap();

        group.bench_with_input(BenchmarkId::new("array", i), &i, |b, input| {
            let mut buffer = [0u8; 8];
            b.iter(|| {
                encode_array(black_box(*input), black_box(&mut buffer));
            });
        });
        group.bench_with_input(BenchmarkId::new("slice", i), &i, |b, input| {
            let mut buffer = [0u8; 8];
            b.iter(|| {
                encode_slice(black_box(*input), black_box(&mut buffer));
            });
        });
        group.bench_with_input(BenchmarkId::new("array_16", i), &i, |b, input| {
            let mut buffer = [0u8; 8 * 16];
            b.iter(|| {
                encode_16_array(black_box(*input), black_box(&mut buffer));
            });
        });
        group.bench_with_input(BenchmarkId::new("slice_16", i), &i, |b, input| {
            let mut buffer = [0u8; 8 * 16];
            b.iter(|| {
                encode_16_slice(black_box(*input), black_box(&mut buffer));
            });
        });
    }
    group.finish();
}

fn decode(c: &mut Criterion) {
    let mut group = c.benchmark_group("varint/decode");
    for i in [0, 1, 5, 6, 13, 14, 29, 30, 61] {
        let i = VarInt::new(2u64.pow(i)).unwrap();

        group.bench_with_input(BenchmarkId::new("array", i), &i, |b, input| {
            let mut buffer = [0; 8];
            encode_array(black_box(*input), black_box(&mut buffer));
            b.iter(|| {
                black_box(decode_array(black_box(&buffer)));
            });
        });
        group.bench_with_input(BenchmarkId::new("slice", i), &i, |b, input| {
            let mut buffer = [0; 8];
            encode_slice(black_box(*input), black_box(&mut buffer));
            b.iter(|| {
                black_box(decode_slice(black_box(&buffer)));
            });
        });
        group.bench_with_input(BenchmarkId::new("array_16", i), &i, |b, input| {
            let mut buffer = [0; 8 * 16];
            encode_16_array(black_box(*input), black_box(&mut buffer));
            b.iter(|| {
                black_box(decode_16_array(black_box(&buffer)));
            });
        });
        group.bench_with_input(BenchmarkId::new("slice_16", i), &i, |b, input| {
            let mut buffer = [0; 8 * 16];
            encode_16_slice(black_box(*input), black_box(&mut buffer));
            b.iter(|| {
                black_box(decode_16_slice(black_box(&buffer)));
            });
        });
    }
    group.finish();
}
