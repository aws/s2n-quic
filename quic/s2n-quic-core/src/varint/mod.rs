// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use core::{
    convert::{TryFrom, TryInto},
    fmt,
    ops::Deref,
};
use s2n_codec::{decoder_value, Encoder, EncoderValue};

#[cfg(any(test, feature = "generator"))]
use bolero_generator::*;

use crate::event::IntoEvent;

mod table;
#[cfg(test)]
mod tests;

//= https://www.rfc-editor.org/rfc/rfc9000#section-16
//# QUIC packets and frames commonly use a variable-length encoding for
//# non-negative integer values.  This encoding ensures that smaller
//# integer values need fewer bytes to encode.

//# The QUIC variable-length integer encoding reserves the two most
//# significant bits of the first byte to encode the base 2 logarithm of
//# the integer encoding length in bytes.  The integer value is encoded
//# on the remaining bits, in network byte order.

pub const MAX_VARINT_VALUE: u64 = 4_611_686_018_427_387_903;

#[derive(Debug)]
pub struct VarIntError;

impl fmt::Display for VarIntError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "varint range exceeded")
    }
}

#[cfg(feature = "std")]
impl std::error::Error for VarIntError {}

// === API ===

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, PartialOrd, Ord)]
#[cfg_attr(any(feature = "generator", test), derive(TypeGenerator))]
pub struct VarInt(#[cfg_attr(any(feature = "generator", test), generator(Self::GENERATOR))] u64);

impl fmt::Display for VarInt {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl VarInt {
    pub const MAX: Self = Self(MAX_VARINT_VALUE);

    pub const ZERO: Self = Self(0);

    #[cfg(any(feature = "generator", test))]
    const GENERATOR: core::ops::RangeInclusive<u64> = 0..=MAX_VARINT_VALUE;

    #[inline(always)]
    pub fn new(v: u64) -> Result<Self, VarIntError> {
        if v > MAX_VARINT_VALUE {
            return Err(VarIntError);
        }
        Ok(Self(v))
    }

    /// Returns a `VarInt` without validating the value is less than VarInt::MAX
    ///
    /// # Safety
    ///
    /// Callers need to ensure the value is less than or equal to VarInt::MAX
    #[inline(always)]
    pub const unsafe fn new_unchecked(value: u64) -> Self {
        Self(value)
    }

    #[inline(always)]
    pub const fn from_u8(v: u8) -> Self {
        Self(v as u64)
    }

    #[inline(always)]
    pub const fn from_u16(v: u16) -> Self {
        Self(v as u64)
    }

    #[inline(always)]
    pub const fn from_u32(v: u32) -> Self {
        Self(v as u64)
    }

    #[inline(always)]
    pub const fn as_u64(self) -> u64 {
        self.0
    }

    #[inline]
    pub fn checked_add(self, value: Self) -> Option<Self> {
        Self::new(self.0.checked_add(value.0)?).ok()
    }

    #[inline]
    pub fn checked_add_usize(self, value: usize) -> Option<Self> {
        let value = value.try_into().ok()?;
        self.checked_add(value)
    }

    #[inline]
    #[must_use]
    pub fn saturating_add(self, value: Self) -> Self {
        Self::new(self.0.saturating_add(value.0)).unwrap_or(Self::MAX)
    }

    #[inline]
    pub fn checked_sub(self, value: Self) -> Option<Self> {
        Some(Self(self.0.checked_sub(value.0)?))
    }

    #[inline]
    #[must_use]
    pub fn saturating_sub(self, value: Self) -> Self {
        Self(self.0.saturating_sub(value.0))
    }

    #[inline]
    pub fn checked_mul(self, value: Self) -> Option<Self> {
        Self::new(self.0.checked_mul(value.0)?).ok()
    }

    #[inline]
    #[must_use]
    pub fn saturating_mul(self, value: Self) -> Self {
        Self::new(self.0.saturating_mul(value.0)).unwrap_or(Self::MAX)
    }

    #[inline]
    pub fn checked_div(self, value: Self) -> Option<Self> {
        Some(Self(self.0.checked_div(value.0)?))
    }

    /// Re-encodes a replacement value where `self` was used as a placeholder.
    #[inline]
    pub fn encode_updated<E: Encoder>(self, replacement: Self, encoder: &mut E) {
        debug_assert!(
            self.table_entry().len >= replacement.table_entry().len,
            "the replacement encoding_size should not be greater than the previous value"
        );

        // don't use the basic version to avoid overwriting things
        self.table_entry()
            .format(replacement.0)
            .encode_maybe_undersized(encoder)
    }

    #[inline(always)]
    fn table_entry(self) -> table::Entry {
        table::Entry::read(self.0)
    }

    #[inline(always)]
    fn format(self) -> table::Formatted {
        table::Formatted::new(self.0)
    }
}

impl EncoderValue for VarInt {
    #[inline(always)]
    fn encode<E: Encoder>(&self, encoder: &mut E) {
        self.format().encode(encoder);
    }

    #[inline(always)]
    fn encoding_size(&self) -> usize {
        self.format().encoding_size()
    }

    #[inline(always)]
    fn encoding_size_for_encoder<E: Encoder>(&self, encoder: &E) -> usize {
        self.format().encoding_size_for_encoder(encoder)
    }
}

impl IntoEvent<u64> for VarInt {
    #[inline]
    fn into_event(self) -> u64 {
        self.as_u64()
    }
}

decoder_value!(
    impl<'a> VarInt {
        fn decode(buffer: Buffer) -> Result<Self> {
            let header = buffer.peek_byte(0)?;

            Ok(match (header >> 6) & 0b11 {
                0b00 => {
                    let value = header & (2u8.pow(6) - 1);
                    let buffer = buffer.skip(1)?;
                    (Self(value.into()), buffer)
                }
                0b01 => {
                    let (value, buffer) = buffer.decode::<u16>()?;
                    let value = value & (2u16.pow(14) - 1);
                    (Self(value.into()), buffer)
                }
                0b10 => {
                    let (value, buffer) = buffer.decode::<u32>()?;
                    let value = value & (2u32.pow(30) - 1);
                    (Self(value.into()), buffer)
                }
                0b11 => {
                    let (value, buffer) = buffer.decode::<u64>()?;
                    let value = value & (2u64.pow(62) - 1);
                    (Self(value), buffer)
                }
                _ => unsafe { core::hint::unreachable_unchecked() },
            })
        }
    }
);

impl AsRef<u64> for VarInt {
    #[inline]
    fn as_ref(&self) -> &u64 {
        &self.0
    }
}

impl Deref for VarInt {
    type Target = u64;

    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

macro_rules! impl_from_lesser {
    ($ty:ty) => {
        impl From<$ty> for VarInt {
            #[inline]
            fn from(value: $ty) -> Self {
                Self(value.into())
            }
        }

        impl TryInto<$ty> for VarInt {
            type Error = <$ty as TryFrom<u64>>::Error;

            #[inline]
            fn try_into(self) -> Result<$ty, Self::Error> {
                self.0.try_into()
            }
        }
    };
}

impl_from_lesser!(u8);
impl_from_lesser!(u16);
impl_from_lesser!(u32);

impl From<VarInt> for u64 {
    #[inline]
    fn from(v: VarInt) -> u64 {
        v.0
    }
}

impl TryFrom<usize> for VarInt {
    type Error = VarIntError;

    #[inline]
    fn try_from(value: usize) -> Result<Self, Self::Error> {
        Self::new(value as u64)
    }
}

impl TryInto<usize> for VarInt {
    type Error = <usize as TryFrom<u64>>::Error;

    #[inline]
    fn try_into(self) -> Result<usize, Self::Error> {
        self.0.try_into()
    }
}

impl TryFrom<u64> for VarInt {
    type Error = VarIntError;

    #[inline]
    fn try_from(value: u64) -> Result<Self, Self::Error> {
        Self::new(value)
    }
}

impl TryFrom<u128> for VarInt {
    type Error = VarIntError;

    #[inline]
    fn try_from(value: u128) -> Result<Self, Self::Error> {
        if value > MAX_VARINT_VALUE as u128 {
            Err(VarIntError)
        } else {
            Ok(Self(value as u64))
        }
    }
}

impl core::ops::Add for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn add(self, rhs: Self) -> Self {
        if cfg!(debug_assertions) {
            self.checked_add(rhs).expect("VarInt overflow occurred")
        } else {
            Self(self.0 + rhs.0)
        }
    }
}

impl core::ops::Add<usize> for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn add(self, rhs: usize) -> Self {
        if cfg!(debug_assertions) {
            self.checked_add(VarInt::new(rhs as u64).expect("VarInt overflow occurred"))
                .expect("VarInt overflow occurred")
        } else {
            Self(self.0 + rhs as u64)
        }
    }
}

impl core::ops::AddAssign<Self> for VarInt {
    #[inline]
    #[track_caller]
    fn add_assign(&mut self, rhs: Self) {
        if cfg!(debug_assertions) {
            *self = self.checked_add(rhs).expect("VarInt overflow occurred")
        } else {
            self.0 += rhs.0
        }
    }
}

impl core::ops::AddAssign<usize> for VarInt {
    #[inline]
    #[track_caller]
    fn add_assign(&mut self, rhs: usize) {
        if cfg!(debug_assertions) {
            *self = self
                .checked_add(VarInt::new(rhs as u64).expect("VarInt overflow occurred"))
                .expect("VarInt overflow occurred")
        } else {
            self.0 += rhs as u64
        }
    }
}

impl core::ops::Sub for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn sub(self, rhs: Self) -> Self {
        // Bounds check is inherited from u64
        Self(self.0 - rhs.0)
    }
}

impl core::ops::Sub<usize> for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn sub(self, rhs: usize) -> Self {
        // Bounds check is inherited from u64
        Self(self.0 - rhs as u64)
    }
}

impl core::ops::SubAssign<Self> for VarInt {
    #[inline]
    #[track_caller]
    fn sub_assign(&mut self, rhs: Self) {
        // Bounds check is inherited from u64
        self.0 -= rhs.0
    }
}

impl core::ops::SubAssign<usize> for VarInt {
    #[inline]
    #[track_caller]
    fn sub_assign(&mut self, rhs: usize) {
        // Bounds check is inherited from u64
        self.0 -= rhs as u64
    }
}

impl core::ops::Mul for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn mul(self, rhs: Self) -> Self {
        if cfg!(debug_assertions) {
            self.checked_mul(rhs).expect("VarInt overflow occurred")
        } else {
            Self(self.0 * rhs.0)
        }
    }
}

impl core::ops::Mul<usize> for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn mul(self, rhs: usize) -> Self {
        if cfg!(debug_assertions) {
            self.checked_mul(VarInt::new(rhs as u64).expect("VarInt overflow occurred"))
                .expect("VarInt overflow occurred")
        } else {
            Self(self.0 * rhs as u64)
        }
    }
}

impl core::ops::MulAssign<Self> for VarInt {
    #[inline]
    #[track_caller]
    fn mul_assign(&mut self, rhs: Self) {
        if cfg!(debug_assertions) {
            *self = self.checked_mul(rhs).expect("VarInt overflow occurred")
        } else {
            self.0 *= rhs.0
        }
    }
}

impl core::ops::MulAssign<usize> for VarInt {
    #[inline]
    #[track_caller]
    fn mul_assign(&mut self, rhs: usize) {
        if cfg!(debug_assertions) {
            *self = self
                .checked_mul(VarInt::new(rhs as u64).expect("VarInt overflow occurred"))
                .expect("VarInt overflow occurred")
        } else {
            self.0 *= rhs as u64
        }
    }
}

impl core::ops::Div for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn div(self, rhs: Self) -> Self {
        // Bounds check is inherited from u64
        Self(self.0 / rhs.0)
    }
}

impl core::ops::Div<usize> for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn div(self, rhs: usize) -> Self {
        // Bounds check is inherited from u64
        Self(self.0 / rhs as u64)
    }
}

impl core::ops::DivAssign<Self> for VarInt {
    #[inline]
    #[track_caller]
    fn div_assign(&mut self, rhs: Self) {
        // Bounds check is inherited from u64
        self.0 /= rhs.0
    }
}

impl core::ops::DivAssign<usize> for VarInt {
    #[inline]
    #[track_caller]
    fn div_assign(&mut self, rhs: usize) {
        // Bounds check is inherited from u64
        self.0 /= rhs as u64
    }
}

impl core::ops::Rem for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn rem(self, rhs: Self) -> Self {
        // Bounds check is inherited from u64
        Self(self.0.rem(rhs.0))
    }
}

impl core::ops::Rem<usize> for VarInt {
    type Output = Self;

    #[inline]
    #[track_caller]
    fn rem(self, rhs: usize) -> Self {
        // Bounds check is inherited from u64
        Self(self.0.rem(rhs as u64))
    }
}

impl core::ops::RemAssign<Self> for VarInt {
    #[inline]
    #[track_caller]
    fn rem_assign(&mut self, rhs: Self) {
        // Bounds check is inherited from u64
        self.0 %= rhs.0
    }
}

impl core::ops::RemAssign<usize> for VarInt {
    #[inline]
    #[track_caller]
    fn rem_assign(&mut self, rhs: usize) {
        // Bounds check is inherited from u64
        self.0 %= rhs as u64
    }
}

impl PartialEq<u64> for VarInt {
    #[inline]
    fn eq(&self, other: &u64) -> bool {
        self.0.eq(other)
    }
}

impl PartialEq<usize> for VarInt {
    #[inline]
    fn eq(&self, other: &usize) -> bool {
        self.0.eq(&(*other as u64))
    }
}

impl PartialOrd<u64> for VarInt {
    #[inline]
    fn partial_cmp(&self, other: &u64) -> Option<core::cmp::Ordering> {
        self.0.partial_cmp(other)
    }
}

impl PartialOrd<usize> for VarInt {
    #[inline]
    fn partial_cmp(&self, other: &usize) -> Option<core::cmp::Ordering> {
        self.0.partial_cmp(&(*other as u64))
    }
}
