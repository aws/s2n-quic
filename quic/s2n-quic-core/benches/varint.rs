// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use criterion::{black_box, criterion_group, criterion_main, Criterion};
use s2n_codec::{DecoderBuffer, EncoderBuffer, EncoderValue};
use s2n_quic_core::varint::VarInt;

fn round_trip(c: &mut Criterion) {
    for i in [0, 1, 5, 6, 13, 14, 29, 30, 61].iter() {
        c.bench_function(&format!("round trip 2^{}", i), move |b| {
            let expected = black_box(VarInt::new(2u64.pow(*i)).unwrap());
            let mut buffer = vec![0; 8];

            b.iter(move || {
                expected.encode(&mut EncoderBuffer::new(&mut buffer));
                let (actual, _) = DecoderBuffer::new(&buffer).decode::<VarInt>().unwrap();
                assert_eq!(actual, expected);
            })
        });
    }
}

criterion_group!(benches, round_trip);
criterion_main!(benches);
