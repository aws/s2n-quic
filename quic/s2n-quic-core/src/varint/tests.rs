// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use crate::varint::VarInt;
use bolero::check;
use core::mem::size_of;
use s2n_codec::{
    assert_codec_round_trip_bytes, assert_codec_round_trip_value, DecoderBuffer, EncoderBuffer,
};

#[test]
#[cfg_attr(miri, ignore)] // This test is too expensive for miri to complete in a reasonable amount of time
fn round_trip_bytes_test() {
    check!().for_each(|input| {
        assert_codec_round_trip_bytes!(VarInt, input);
    });
}

#[test]
#[cfg_attr(kani, kani::proof, kani::unwind(10), kani::solver(cadical))]
fn round_trip_values_test() {
    check!().with_type().cloned().for_each(|v| {
        if let Ok(v) = VarInt::new(v) {
            assert_codec_round_trip_value!(VarInt, v);
        } else {
            assert!(v > MAX_VARINT_VALUE);
        }
    })
}

#[test]
#[cfg_attr(miri, ignore)] // snapshot tests don't work on miri
fn table_snapshot_test() {
    use insta::assert_debug_snapshot;
    assert_debug_snapshot!("max_value", MAX_VARINT_VALUE);

    // These values are derived from the "usable bits" column in the table: V and V-1
    for i in [0, 1, 5, 6, 13, 14, 29, 30, 61] {
        assert_debug_snapshot!(
            format!("table_2_pow_{}_", i),
            table::Entry::read(2u64.pow(i))
        );
    }
}

//= https://www.rfc-editor.org/rfc/rfc9000#appendix-A.1
//# For example, the eight-byte sequence 0xc2197c5eff14e88c decodes to
//# the decimal value 151,288,809,941,952,652; the four-byte sequence
//# 0x9d7f3e7d decodes to 494,878,333; the two-byte sequence 0x7bbd
//# decodes to 15,293; and the single byte 0x25 decodes to 37 (as does
//# the two-byte sequence 0x4025).

macro_rules! sequence_test {
    ($name:ident($input:expr, $expected:expr)) => {
        #[test]
        fn $name() {
            use s2n_codec::assert_codec_round_trip_value;

            let input = $input;
            let expected = VarInt::new($expected).unwrap();
            let actual_bytes = assert_codec_round_trip_value!(VarInt, expected);
            assert_eq!(&input[..], &actual_bytes[..]);
        }
    };
}

sequence_test!(eight_byte_sequence_example(
    [0xc2, 0x19, 0x7c, 0x5e, 0xff, 0x14, 0xe8, 0x8c],
    151_288_809_941_952_652
));

sequence_test!(four_byte_sequence_example(
    [0x9d, 0x7f, 0x3e, 0x7d],
    494_878_333
));

sequence_test!(two_byte_sequence_example([0x7b, 0xbd], 15293));

sequence_test!(one_byte_sequence_example([0x25], 37));

//= https://www.rfc-editor.org/rfc/rfc9000#section-16
//= type=test
// # +======+========+=============+=======================+
// # | 2MSB | Length | Usable Bits | Range                 |
// # +======+========+=============+=======================+
// # | 00   | 1      | 6           | 0-63                  |
// # +------+--------+-------------+-----------------------+
// # | 01   | 2      | 14          | 0-16383               |
// # +------+--------+-------------+-----------------------+
// # | 10   | 4      | 30          | 0-1073741823          |
// # +------+--------+-------------+-----------------------+
// # | 11   | 8      | 62          | 0-4611686018427387903 |
// # +------+--------+-------------+-----------------------+
#[test]
#[cfg_attr(miri, ignore)]
#[cfg_attr(kani, kani::proof, kani::unwind(3), kani::solver(cadical))]
fn one_byte_sequence_test() {
    bolero::check!()
        .with_type()
        .cloned()
        .for_each(|mut first_byte: u8| {
            first_byte &= 0x3f;
            let byte_sequence = [first_byte];
            let expected = VarInt::new(first_byte as u64).unwrap();
            let actual_bytes = assert_codec_round_trip_value!(VarInt, expected);
            assert_eq!(&byte_sequence[..], &actual_bytes[..]);
        });
}

#[test]
#[cfg_attr(miri, ignore)]
#[cfg_attr(kani, kani::proof, kani::unwind(4), kani::solver(cadical))]
fn two_byte_sequence_test() {
    // the s2n-quic implementation always chooses the smallest encoding possible.
    // this means if we wish to test two-byte sequences, we need to encode a number
    // that is > 63.
    let g = gen::<(u8, u8)>().filter_gen(|(_, b)| *b > 63);

    bolero::check!()
        .with_generator(g)
        .cloned()
        .for_each(|(mut first_byte, second_byte)| {
            first_byte = (first_byte & 0x3f) | 0x40;
            let byte_sequence = [first_byte, second_byte];
            let expected_val = u16::from_be_bytes([first_byte & 0x3f, second_byte]);
            assert!(expected_val <= 16383);
            let expected = VarInt::new(expected_val as u64).unwrap();
            let actual_bytes = assert_codec_round_trip_value!(VarInt, expected);
            assert_eq!(&byte_sequence[..], &actual_bytes[..]);
        });
}

#[test]
#[cfg_attr(miri, ignore)]
#[cfg_attr(kani, kani::proof, kani::unwind(10), kani::solver(cadical))]
fn four_byte_sequence_test() {
    // The s2n-quic implementation always chooses the smallest encoding possible.
    // This means if we wish to test the four-byte sequences, we need to encode a number
    // that is > 16383 or 0b0011 1111 1111 1111.
    let g = gen::<(u8, u8, u8)>().filter_gen(|(_, _, c)| *c > 0x3f);

    bolero::check!().with_generator(g).cloned().for_each(
        |(mut first_byte, second_byte, third_byte)| {
            first_byte = (first_byte & 0x3f) | 0x80;
            let byte_sequence = [first_byte, second_byte, third_byte, 0xff];
            let expected_val: u32 =
                u32::from_be_bytes([first_byte & 0x3f, second_byte, third_byte, 0xff]);
            assert!(expected_val <= 1073741823);
            let expected = VarInt::new(expected_val as u64).unwrap();
            let actual_bytes = assert_codec_round_trip_value!(VarInt, expected);
            assert_eq!(&byte_sequence[..], &actual_bytes[..]);
        },
    );
}

#[test]
#[cfg_attr(miri, ignore)]
#[cfg_attr(kani, kani::proof, kani::unwind(10), kani::solver(cadical))]
fn eight_byte_sequence_test() {
    // The s2n-quic implementation always chooses the smallest encoding possible.
    // This means if we wish to test eight-byte sequences, we need to encode a number
    // that is > 1073741823 or 0b0011 1111 1111 1111 1111 1111 1111 1111
    let g = gen::<(u8, u8, u8, u8)>().filter_gen(|(_, _, _, d)| *d > 0x3f);

    bolero::check!().with_generator(g).cloned().for_each(
        |(mut first_byte, second_byte, third_byte, fourth_byte)| {
            first_byte = (first_byte & 0x3f) | 0xc0;
            let byte_sequence = [
                first_byte,
                second_byte,
                third_byte,
                fourth_byte,
                0xff,
                0xff,
                0xff,
                0xff,
            ];
            let expected_val: u64 = u64::from_be_bytes([
                first_byte & 0x3f,
                second_byte,
                third_byte,
                fourth_byte,
                0xff,
                0xff,
                0xff,
                0xff,
            ]);
            assert!(expected_val <= 4611686018427387903);
            let expected = VarInt::new(expected_val).unwrap();
            let actual_bytes = assert_codec_round_trip_value!(VarInt, expected);
            assert_eq!(&byte_sequence[..], &actual_bytes[..]);
        },
    );
}

fn test_update(initial: VarInt, expected: VarInt, encoder: &mut EncoderBuffer) {
    encoder.set_position(0);
    initial.encode_updated(expected, encoder);
    let decoder = DecoderBuffer::new(encoder.as_mut_slice());
    let (actual, _) = decoder.decode::<VarInt>().unwrap();
    assert_eq!(expected, actual);
}

#[test]
fn encode_updated_test() {
    let mut buffer = [0u8; size_of::<VarInt>()];
    let mut encoder = EncoderBuffer::new(&mut buffer);
    let initial = VarInt::from_u16(1 << 14);
    encoder.encode(&initial);

    test_update(initial, VarInt::from_u32(0), &mut encoder);
    test_update(initial, VarInt::from_u32(1 << 14), &mut encoder);
    test_update(initial, VarInt::from_u32(1 << 29), &mut encoder);
}

#[test]
#[should_panic]
fn encode_updated_invalid_test() {
    let mut buffer = [0u8; size_of::<VarInt>()];
    let mut encoder = EncoderBuffer::new(&mut buffer);
    let initial = VarInt::from_u16(1 << 14);
    encoder.encode(&initial);

    test_update(initial, VarInt::from_u32(1 << 30), &mut encoder);
}

#[test]
#[cfg_attr(kani, kani::proof, kani::unwind(5), kani::solver(kissat))]
fn table_differential_test() {
    //= https://www.rfc-editor.org/rfc/rfc9000#section-16
    //= type=test
    //# This means that integers are encoded on 1, 2, 4, or 8 bytes and can
    //# encode 6-, 14-, 30-, or 62-bit values, respectively.  Table 4
    //# summarizes the encoding properties.
    //#
    //#        +======+========+=============+=======================+
    //#        | 2MSB | Length | Usable Bits | Range                 |
    //#        +======+========+=============+=======================+
    //#        | 00   | 1      | 6           | 0-63                  |
    //#        +------+--------+-------------+-----------------------+
    //#        | 01   | 2      | 14          | 0-16383               |
    //#        +------+--------+-------------+-----------------------+
    //#        | 10   | 4      | 30          | 0-1073741823          |
    //#        +------+--------+-------------+-----------------------+
    //#        | 11   | 8      | 62          | 0-4611686018427387903 |
    //#        +------+--------+-------------+-----------------------+

    check!()
        .with_generator(0..=MAX_VARINT_VALUE)
        .cloned()
        .for_each(|v| {
            let expected = table::Entry::read_rfc(v);
            let actual = table::Entry::read_optimized(v);
            assert_eq!(actual, expected);
        })
}

#[test]
#[cfg_attr(kani, kani::proof, kani::unwind(5), kani::solver(kissat))]
fn checked_ops_test() {
    check!().with_type().cloned().for_each(|(a, b)| {
        if let (Ok(a_v), Ok(b_v)) = (VarInt::new(a), VarInt::new(b)) {
            // compares checked operations against the underlying u64 operations
            macro_rules! checked {
                ($checked:ident, $op:tt) => {{
                    if let Some(res) = a_v.$checked(b_v) {
                        // make sure the behavior matches that of u64
                        assert_eq!(Some(res.as_u64()), a.$checked(b));

                        // make sure the normal operation works as well
                        assert_eq!(res, a_v $op b_v);
                    } else {
                        // make sure the behavior matches that of u64
                        let expected = a.$checked(b).and_then(|v| VarInt::new(v).ok());
                        assert!(expected.is_none());
                    }
                }};
            }

            checked!(checked_add, +);
            checked!(checked_sub, -);

            // kani currently spins for a really long time to handle these checks
            #[cfg(any(not(kani), kani_slow))]
            checked!(checked_mul, *);
            #[cfg(any(not(kani), kani_slow))]
            checked!(checked_div, /);

            // compares saturating operations against the underlying u64 operations
            macro_rules! saturating {
                ($name:ident) => {{
                    let actual = a_v.$name(b_v);
                    let expected = VarInt::new(a.$name(b)).unwrap_or(VarInt::MAX);
                    assert_eq!(actual, expected);
                }};
            }

            saturating!(saturating_add);

            // kani currently spins for a really long time to handle these checks
            #[cfg(any(not(kani), kani_slow))]
            saturating!(saturating_sub);
            #[cfg(any(not(kani), kani_slow))]
            saturating!(saturating_mul);
        }
    })
}
