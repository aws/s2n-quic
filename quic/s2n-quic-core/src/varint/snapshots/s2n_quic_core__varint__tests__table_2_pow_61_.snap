---
source: quic/s2n-quic-core/src/varint/tests.rs
expression: "table::Entry::read(2u64.pow(i))"
---
Entry {
    two_bit: 3,
    two_bit_be: [
        192,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
    ],
    len: 8,
    usable_bits: 62,
    shift: 0,
}
